File: R/send.inat.R
send.inat <- function(obs, id=NULL, conda=NULL, python=NULL) {
  # obs should be a data.frame with some mandatory fields:
  # - Species: binomial species name
  # - Date: date of observation
  # - Lat and Long: geographical coordinates of observation
  # - Precision: precision of observation coordinates
  # - Path: path to image associated to observation
  
  # id should be a data.frame including the following information
  # - user: user account name in iNaturalist
  # - pwd: password associated to user account name in iNaturalist
  # 
  
  require(reticulate)
  
  if(is.null(id)) 
    stop("Missing identification data")
  
  if(!is.data.frame(id))
    stop("id must be a data.frame")
  
  if(any(!c("user","pwd","app.id","app.pwd")%in%names(id)))
     stop("Incorrect identification data")
  
  if(!is.data.frame(obs))
    stop("obs must be a data.frame")
  
  if(any(!c("Species", "Date", "Lat", "Long", "Path", "Precision")%in%names(obs)))
    stop("Incorrect fields in obs")
  
  # See POST  options in https://www.inaturalist.org/pages/api+reference
  # https://pythonawesome.com/an-inaturalist-api-client-for-python/
  
  # Some previous (unsucessful) tries
  # virtualenv_install(envname = "send.inat",
  #                               packages="pyinaturalist")
  # use_virtualenv(virtualenv = "send.inat")
  
  if(!is.null(conda)) {
    use_condaenv(condaenv=conda)
  } 
  
  if(is.null(conda) & !is.null(python)) {
    use_python(python=python)
  }
  
  if(is.null(conda) & !is.null(python)) {
    use_virtualenv("__main__")
    virtualenv_install(envname="__main__", packages ='pyinaturalist')
  }
  
  py_run_string(paste("from pyinaturalist import *\ntoken = get_access_token(username='",
                                  id$user,
                                  "',\npassword='",
                                  id$pwd,
                                  "',\napp_id='",
                                  id$app.id,
                                  "',\napp_secret='",
                                  id$app.pwd,
              "')\ntaxon = get_taxa(q='",
                                  obs$Species,
                                  "', rank=['species'])\nfrom datetime import datetime\nresponse = create_observation(taxon_id=taxon['results'][0]['id'],observed_on_string='",
                                  obs$Date,
                                  "',\nlatitude=",
                                  obs$Lat,
                                  ",\nlongitude=",
                                  obs$Long,
                                  ",\npositional_accuracy=",
                                  obs$Precision,
                                  ",\naccess_token=token)\nnew_observation_id = response['id']\nupload_photos(new_observation_id,'",
                                  obs$Path,
              "', access_token=token)\n", sep=""),
        convert = T)
}

File: send.inat.py
# Needed libraries must be installed
pip install pandas
# conda install pandas
pip install pyinaturalist

import pandas as pd
from pyinaturalist import *
import os

from pyinaturalist.rest_api import add_photo_to_observation

os.chdir('C:/Users/munozfra/Documents/Rprojects/naturalistDB')

# The ID file should contain the identification information
id=pd.read_csv (r'id.csv',sep=';')
# Or directly include below
#{'user':['XXX'],
#'pwd':['XXX'],
#'app.id':['XXX'],
#'app.pwd':['XXX']}
#id=pd.DataFrame(id)

df = pd.read_csv (r'20220812 Export iNaturalist.csv',sep=';',decimal=',',encoding='latin-1')
print (df)
# Get values
df.values[1][0]
df['Species']
df.get('Species)[1]

# Add file path information
df2 = pd.DataFrame().assign(Path='C:/PlantNet/'+df['File.name'])

token = get_access_token(username=id.get('user')[0],password=id.get('pwd')[0],
app_id=id.get('app.id')[0],app_secret=id.get('app.pwd')[0])

for index in range(21, 40): #int(df.shape[0])):
  
  precision=10000
  
  taxon = get_taxa(q=df.get('Taxon')[index],rank=['species'])
  
  from datetime import datetime
  
  response = create_observation(taxon_id=taxon['results'][0]['id'],
  observed_on_string=df.get('Time')[index],
  latitude=df.get('lat')[index],
  longitude=df.get('long')[index],
  positional_accuracy=precision,
  description=df2.get('Path')[index],
  access_token=token)
  
  new_observation_id = response['id']
  
  # Alternative options to send pictures
  # upload_photos(new_observation_id,df2.get('Path')[index], access_token=token)
  try:
    add_photo_to_observation ( observation_id=new_observation_id, access_token=token,
  file_object=df2.get('Path')[index], photos=df2.get('Path')[index]) 
  except:
    print(index)
    
# Next: try to add picture to an existing observation without photo
